### PR TITLE
#176408984 make enquiries support vendor

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Make enquiries support vendor

#### Description of Task to be completed?
Make enquiries support vendor

#### What are the relevant pivotal tracker stories?
#176408984

### CODE REVIEW

File: test/controllers/enquiry.controller.test.js [hunk 1]
```diff
@@ -521,7 +531,7 @@ describe('Enquiry Controller', () => {
       it('returns a approved enquiry', (done) => {
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:531]:
```suggestion
      it('returns an approved enquiry', (done) => {
```

File: test/controllers/enquiry.controller.test.js [hunk 2]
```diff
@@ -659,104 +669,138 @@ describe('Enquiry Controller', () => {
...
+          useExistingUser: true,
+        }),
+      );
     });
 
     describe('when enquiries exist in db', () => {
       beforeEach(async () => {
+        await Property.insertMany(dummyProperties);
+        await Enquiry.insertMany(userEnquiries);
         await addEnquiry(enquiry);
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:742]: Not sure why you left this. I believe you can remove that line and use the custom pagination test 🤷🏽‍♂️

File: test/controllers/enquiry.controller.test.js [hunk 3]
```diff
@@ -659,104 +669,138 @@ describe('Enquiry Controller', () => {
...
+      context('when request is sent by admin token', () => {
+        it('returns all enquiries', (done) => {
           request()
-            .get('/api/v1/enquiry/all')
+            [method](endpoint)
             .set('authorization', adminToken)
             .end((err, res) => {
               expect(res).to.have.status(200);
               expect(res.body.success).to.be.eql(true);
-              expect(res.body).to.have.property('enquiries');
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:699]: Can you reintroduce this tests back as the new test doesn't replace it. In testing an endpoint with pagination, lets try to keep the following in mind. In this case, we have admin, user and vendor. 

Based on the information on the comment section, here is the breakdown
1. Test when nothing is in the database for the three  (already done on line 727)
2. Test that the pagination returns the right value for the three 
   - It is preferable to use the custom pagination test as it contains multiple test
   - A recurring mistake I notice is that you usually try to force the test in a single test assertion (i.e a describe/context block), it is not compulsory to do this. You might split the test into different blocks to make it more readable and better organized.
   - user test for this seems missing
3. Test that the right value is returned. 
    - The pagination test only deals with the numbers returned, not the validity of the data returned, so the pagination test cannot replace this
    - You will need to reintroduce this test you removed.
4. Add more tests based on the scenarios you are testing. Some of this can be incorporated in the pagination test setup:
   - You added some here like testing that vendor2 cannot access vendor1 enquiries
   - You can add more like ensuring that a user can only access his enquiry
   - Admin should be able to view all enquiries (This can be incorporated in the pagination test)

File: server/services/enquiry.service.js [hunk 4]
```diff
@@ -89,20 +104,21 @@ export const getEnquiry = async (enquiryId) =>
...
         'propertyInfo.assignedTo': 0,
-        'approvedBy.assignedProperties': 0,
-        'approvedBy.password': 0,
-        'approvedBy.referralCode': 0,
       },
     },
-  ]);
+  ];
+
+  if (user.role === USER_ROLE.VENDOR) {
```
Comment (andela-hpopoola) [inline server/services/enquiry.service.js:114]: Tests should be written to validate that this is works as expected.

File: server/services/enquiry.service.js [hunk 5]
```diff
@@ -89,20 +104,21 @@ export const getEnquiry = async (enquiryId) =>
...
+    enquiryOptions.unshift({ $match: { vendorId: ObjectId(user._id) } });
+  }
+  if (user.role === USER_ROLE.USER) {
```
Comment (andela-hpopoola) [inline server/services/enquiry.service.js:117]: Tests should be written to validate that this works too.

File: server/models/enquiry.model.js [hunk 6]
```diff
@@ -86,6 +86,10 @@ const EnquirySchema = new mongoose.Schema(
       type: ObjectId,
       required: true,
     },
+    vendorId: {
```
Comment (andela-hpopoola) [inline server/models/enquiry.model.js:89]: I believe this should be added to `test/factories/enquiry.factory.js` too since it is required to make it easier to generate test objects for enquiries.

File: server/services/enquiry.service.js [hunk 7]
```diff
@@ -27,31 +29,35 @@ export const addEnquiry = async (enquiry) => {
...
 
-export const approveEnquiry = async (approvedEnquiry) => {
-  const enquiry = await getEnquiryById(approvedEnquiry.enquiryId).catch((error) => {
+export const approveEnquiry = async ({ enquiryId, vendor }) => {
+  const enquiry = await getEnquiryById(enquiryId).catch((error) => {
     throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
   });
 
   try {
+    if (vendor.role === USER_ROLE.VENDOR && vendor._id.toString() !== enquiry.vendorId.toString()) {
```
Comment (andela-hpopoola) [inline server/services/enquiry.service.js:45]: We might need to add a test to ensure that this works as expected too.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Awesome. You did a great job here.

Thank you for working on this. The implementation makes sense but I am not satisfied with the tests as some of them are not comprehensive enough.

Can you try to always follow this implementation when writing tests for pagination: Please read this thoroughly to ensure that you understand them and always use them as reference for pagination tests.

1. Test when nothing is in the database (when database table contains no data)
2. Test that the pagination returns the right value (use pagination test 99% of the time, the remaining 1% is for edge case)
3. Test that the right values is returned.  The pagination test only deals with the numbers returned, not the validity of the data returned, so the pagination test cannot replace this

4. Add more customized tests based on the scenarios you are testing. These are tests directly related to the code. Examples are 
    - Testing for Valid user
    - Ensuring that vendor2 doesn't have access to vendor 1 information
    - Ensuring that user2 gets his own message
    - Testing for all if statements used in code with the else part
    - E.t.c.

File: test/controllers/enquiry.controller.test.js (hunk 2 above)
Comment (tobieeniafe) [reply test/controllers/enquiry.controller.test.js:742]: this was added to be sure that the user and admin do not return the same data. the user token uses the pagination tests

File: server/models/enquiry.model.js (hunk 6 above)
Comment (tobieeniafe) [reply server/models/enquiry.model.js:89]: this is added automatically in service. gotten from id of the owner of the property

File: test/controllers/enquiry.controller.test.js [hunk 8]
```diff
@@ -510,6 +510,17 @@ describe('Enquiry Controller', () => {
   });
 
   describe('Approve Enquiry', () => {
+    let vendor2Token;
+    const vendor2 = UserFactory.build(
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:514]: Lets make this `vendorUser2` to make it consistent with `vendorUser`.

File: test/controllers/enquiry.controller.test.js [hunk 9]
```diff
@@ -606,22 +656,132 @@ describe('Enquiry Controller', () => {
...
+              expect(res.body.message).to.be.eql('Enquiry not found');
+              done();
+            });
+        });
+      });
+    });
+
+    context('with a token with valid access', () => {
+      // eslint-disable-next-line array-callback-return
+      [demoUser, vendor2].map((user) => {
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:719]: Instead of `demoUser`, lets use something like `regularUser2` to make it easier to relate it.  It will even make more sense to add if you can add `1` to the the first one too.

That way, `regularUser1` makes an enquiry on `vendorUser1` property called `enquiry1` and `regularUser2` makes an enquiry on `vendorUser2` called `enquiry2`. This will make it easier to understand the user, vendor and enquiry you are referring to without stress.

Reading through the code was kinda of difficult to relate with because of the naming convention. I had to revisit the variable definition to understand what they are doing.

File: test/controllers/enquiry.controller.test.js [hunk 10]
```diff
@@ -606,22 +656,132 @@ describe('Enquiry Controller', () => {
...
+              expect(res.body).to.have.property('enquiry');
+              expect(res.body.enquiry._id).to.be.eql(vendor2Enquiry._id.toString());
+              done();
+            });
+        });
+      });
+    });
+
+    context('with a token with invalid access', () => {
+      // eslint-disable-next-line array-callback-return
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:742]: Lets remove this and add `return null` at the end of the test block.

File: test/controllers/enquiry.controller.test.js [hunk 11]
```diff
@@ -606,22 +656,132 @@ describe('Enquiry Controller', () => {
...
+              expect(res.body.success).to.be.eql(true);
+              expect(res.body).to.have.property('enquiry');
+              expect(res.body.enquiry._id).to.be.eql(vendor2Enquiry._id.toString());
+              done();
+            });
+        });
+      });
+    });
+
+    context('with a token with invalid access', () => {
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:741]: The token actually has a valid access but they are not authorized to view another user's enquiry.

```suggestion
    context('when viewing an unauthorized enquiry', () => {
```

File: test/controllers/enquiry.controller.test.js [hunk 12]
```diff
@@ -606,22 +656,132 @@ describe('Enquiry Controller', () => {
...
+              expect(res).to.have.status(404);
+              expect(res.body.success).to.be.eql(false);
+              expect(res.body.message).to.be.eql('Enquiry not found');
+              done();
+            });
+        });
+      });
+    });
+
+    context('with a token with valid access', () => {
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:764]:
```suggestion
	    context('when viewing an authorized enquiry', () => {
```

File: test/controllers/enquiry.controller.test.js [hunk 13]
```diff
@@ -657,7 +817,7 @@ describe('Enquiry Controller', () => {
       it('returns the error', (done) => {
         sinon.stub(Enquiry, 'aggregate').throws(new Error('Type Error'));
         request()
-          .get(`/api/v1/enquiry/${enquiry._id}`)
+          .get(`/api/v1/enquiry/${vendorEnquiry._id}`)
```
Comment (andela-hpopoola) [inline test/controllers/enquiry.controller.test.js:820]: Can we revert the variable back to `enquiry`. `vendorEnquiry` sometimes give the enquiry a different meaning as the enquiry is made by the user. Lets use `enquiry` to be more neutral.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this and implementing the requested changes. Thank you for the response to the comments too, they are really on point 👍🏽 .

I left a few comments on proper naming and removing the eslint comments. I should have approved this but would love you to fix the failing test assertion before approving. Thank you and nice work.

**Review by andela-hpopoola — APPROVED** [review]
Thank you for making the changes. This looks good to me.
